README.md
# cloudSysadmin

A Go program (client) and shell script to automate setting up VMs in the cloud(s) [Vultr]. 

The scripts should be focused on running on FreeBSD. Properly tested on **FreeBSD 13.0**

## Configuration
### ssh Keys
If an ssh key is automatically installed in the server at deployment, to use it from the client:
1. `ssh-add` : add key to keys agent
2. `ssh -i <path_of_key> root@<IP_address>` : explicitly state where the file with the key is located

* Reference: [Vultr Docs - ssh keys](https://www.vultr.com/docs/connect-to-a-server-using-an-ssh-key)

# FreeBSD: General information
## Networking
* [Check open ports on FreeBSD](https://linuxhint.com/check-open-ports-freebsd/): _sockstat_ command

## SSH
* [How to configure ssh key authentication on FreeBSD](https://www.digitalocean.com/community/tutorials/how-to-configure-ssh-key-based-authentication-on-a-freebsd-server)

## MariaDB
* [How to install MariaDB on FreeBSD](https://www.osradar.com/how-to-install-mariadb-on-freebsd-12/)


CHANGELOG.md
## v0.7.2
* Add more specific error/diagnostic message for 401 Unauthorized response.
## v0.7.1
* Require also Vim, when configuring the dotfiles in the VM
## v0.7.0
* Fix big issue (#6), the system now properly sources the packages that will be installed in the VM
## v0.6.0
* Instance's OS is now configurable through the .env file, FreeBSD 13 is still the default option.
## v0.5.0
* Instance's plan is now configurable through the .env file.
## v0.4.0
* Load the configuration details for the server from a .env file
* [BUG] the system does not properly source the packages that have to be installed in the new VM
## v0.3.1
* Fix minor issue: sed was not commenting out the line for _vim-go_ initialization in the vimrc file.
## v0.3.0
* Properly implement deleting an instance into the main.go file
## v0.2.2
* Add unit test for checkResponseAPI()
## v0.2.1
* Add untested deleteInstance method, and refactor code of main.go to enable 'action' flag
## v0.2.0
* Dramatically improve the start-up performance by removing the installation of the _vim-go_ plug from the vimrc file with sed.

## v0.1.0
* First fully automated version. If the API token is placed in the right place, it should all work fully automated.
* [BUG] It still takes to long after the first time it is set up, to get into vim, because _vim-go_ tries to update all required binaries.


vultrClient/cmd/api/helpers_test.go
package main

import (
	"net/http"
	"testing"
)

func TestCheckResponseAPI(t *testing.T) {
	respOK := new(http.Response)
	respOK.StatusCode = 200 // 200 OK

	if err := checkResponseAPI(respOK, 200); err != nil {
		t.Errorf("checkResponseAPI did not recognize 200 OK")
	}

	if err := checkResponseAPI(respOK, 404); err == nil {
		t.Errorf("checkResponseAPI did not recognize different status codes")
	}

}


createVMVultr.sh
#!/bin/sh
# Eduardo Rodriguez [@erodrigufer] 2022 (c) 
# Spawn a VM in Vultr Cloud using a self-made Go client for the Vultr API
#
#########################################################################

#########################################################################
# Default values, just in case that configVM.env does not exist, or it
# even exists, but does not contain these variables. So the program would still
# work with some default values even if the .env file is missing some variables.
HOSTNAME="eee"
LABEL="newInstance"
REGION="fra"
PLAN="vc2-1c-1gb"	# cheapest/most basic plan
OS="447"	# FreeBSD-13
#########################################################################
# If the .env file is configured correctly, its variables should over-write the
# default ones.
if [ -f ./configVM.env ]; then
	. ./configVM.env
fi

# Eliminate credentials file, if it already exists
rm -f ./vm_credentials.secrets
cd ./vultrClient
make build
source  ./secrets/vultrAPI.secrets
cd ..
./vultrClient/build/vultrClient.bin -action "create" -hostname ${HOSTNAME} -label ${LABEL} -sshKey \
	${SSH_KEY} -tokenAPI ${API_TOKEN} -region ${REGION} -plan ${PLAN} -os ${OS} || exit -1

# Source the newly acquired credentials for the VM
source ./vm_credentials.secrets
# Ping the newly created VM every 10 seconds, until it is up and running
ping -o -i 10 ${HOST}
# SSH into the VM, do not ask to check key fingerprinting for new host
# ssh -o "StrictHostKeyChecking no" root@${HOST}
./setupFirstRun.sh 
# After finishing setup, ssh into the VM
ssh ${USER}@${HOST}


setupFirstRun.sh
#!/bin/sh
# Eduardo Rodriguez [@erodrigufer] 2022 (c) 
# This script automates setting up a FreeBSD VM in the cloud after it has been
# successfully spawned.
# - It configures the ecosystem for development; vim, go, dotfiles in general
# - It configures sshd to not close a connection so quickly, and accept public
#	key authentication
# - It installs common useful packages and, if desired, packages for web dev.

# Default values, just in case that configVM.env does not exist, or it
# even exists, but does not contain these variables. So the program would still
# work with some default values even if the .env file is missing some variables.
#########################################################################
# Default user-defined input parameters

MAIN_USER="hap"

# If the .env file is configured correctly, its variables should over-write the
# default ones.
if [ -f ./configVM.env ]; then
	. ./configVM.env
fi

if [ -z ${PACKAGES2INSTALL} ]; then 
# Package naming used by FreeBSD pkg
# lf-26 is the Go terminal file manager
# duf is a modern version of 'df', free disk utility
PACKAGES2INSTALL="git vim curl go lf duf htop"
# hey is used to send load to web applications
fi

if [ -z ${WEBDEVPACKAGES} ]; then 
WEBDEVPACKAGES="mariadb105-server hey sqlite3"
fi

#########################################################################
# Internal global variables, which should not be modified by user

OS=''
# Colors for the text (they do not set the background color)
COLOR_GREEN='\e[0;32m'
NO_COLOR='\033[0m'
COLOR_RED='\033[0;31m'
COLOR_MAGENTA='\033[0;35m'
COLOR_CYAN='\033[0;36m'

SSHD_CONFIG_FILE="./etc/sshd_config"

#########################################################################
# Internal global variables, which can be modified by user

FILE_VM_CREDENTIALS=vm_credentials.secrets # file with ssh credentials for VM
FILE_USER_CONFIG="./configVM.env" # file that describes which packages to 
# install, which instance to initiate and so on...
FILE_NAME=$0 # name of this same file
PATH_IN_VM="/root/" # where to store setup script in VM, from which it is then
					# run inside the VM after establishing ssh connection

#########################################################################
# Helper functions

# Print an info message to stdout. First parameter is string to be logged
print_info(){
	local INFO_COLOR=${COLOR_CYAN}
	printf "[${INFO_COLOR}INFO${NO_COLOR}] $1\n"
}

# Print an error message to stdout. First parameter is string to be logged
print_error(){
	printf "[${COLOR_RED}ERROR${NO_COLOR}] $1\n"
}

#########################################################################

# Check if the script is running in a valid OS (eventual support should be for
# primary FreeBSD and Ubuntu)
check_os(){
	echo "* Checking system distribution..."
	if [ $(uname) = "Linux" ]; then
		# print Linux_Standard_Base release ID use grep in egrep mode (-e), 
		# being case-insensitive (-i) and check for either ubuntu or debian
		# About egrep: with egrep it is easier to write the OR logic ( | ) 
		# otherwise all these characters must be escaped with the normal regex 
		# grammar of grep
		lsb_release -i | grep -i -E "(ubuntu|debian|kali)" || { print_error "System must be debian-based to run script."; exit -1 ; }
		print_error "{$(uname)} System not supported!"	
		OS="debian"
		exit -1
	fi
	if [ $(uname) = "FreeBSD" ]; then
		OS="freebsd"
		return
	fi

	print_error "{$(uname)} System not supported!"	
	exit -1
}

configure_users(){
	# Check if username already exists
	
	adduser ${MAIN_USER}
	# give sudo rights to main user
	adduser ${MAIN_USER} sudo	

}

install_webdev_packages(){
	pkg install --yes ${WEBDEVPACKAGES} && print_info "${WEBDEVPACKAGES} were successfully installed!" || { print_error "Webdev packages installation failed!"; return; }

	# Installing mariaDB on FreeBSD:
	# https://www.osradar.com/how-to-install-mariadb-on-freebsd-12/

	# Enable the mariadb service to start with the system
	sysrc mysql_enable="yes"

	# Start mariadb service
	service mysql-server start

	# Run the mysql configuration script
	# /usr/local/bin/mysql_secure_installation

}

# Install the bare minimum of necessary packages
install_bare_packages(){
	if [ ${OS} = "freebsd" ]; then
		
		# Update the available remote repositories
		pkg update	
		# Install packages without further confirmation (--yes)
		# For more information take a look at 'man pkg-install'
		# To get more info about a particular package run:
		# pkg search -R <name>
		pkg install --yes ${PACKAGES2INSTALL} && print_info "${PACKAGES2INSTALL} were successfully installed!" || { print_error "Packages installation failed!"; }
		
		return
	fi

	if [ ${OS} = "debian" ]; then
		return	
	fi

}

# configure vim, its plugins, tmux, lf, all dotfiles
configure_dotfiles(){
	# If git, vim or curl are not present, return without installing the dotfiles
	which git > /dev/null || return
	which curl > /dev/null || return
	which vim > /dev/null || return
	# Install vim-plug to handle plugins
	# Reference: https://github.com/junegunn/vim-plug
	curl -fLo ~/.vim/autoload/plug.vim --create-dirs \
		    https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim

	local REPO_PATH="/dotfiles"
	# change to home directory
	cd ${HOME}
	git clone https://github.com/erodrigufer/dotfiles.git || { rm -fR ${HOME}${REPO_PATH}; print_error "Cloning GitHub dotfiles repo failed!"; exit -1; }
	cd ${HOME}${REPO_PATH}
	# Eliminate vim-go in the local installation
	# Reference to escape single quotes:
	# https://stackoverflow.com/questions/24509214/how-to-escape-single-quote-in-sed
	sed -i '' -e 's/^Plug.*fatih/"Plug '\''fatih/' ./vimrc
	make || { rm -fR ${HOME}${REPO_PATH}; print_error "Creating symlinks to dotfiles failed!"; exit -1; }

	# Install vim plugins in the background
	vim -E -s +PlugInstall +qall
	
	# we do not need the zshrc file
	rm -f ${HOME}/.zshrc
	
}

# Configure sshd properly, do not disconnect client after short idle time
configure_ssh(){
	# Create a copy (backup) of the sshd original configuration, before changing it
	cp /etc/ssh/sshd_config /etc/ssh/sshd_config.bak

	# Replace original file with new config file
	cp ${PATH_IN_VM}sshd_config /etc/ssh/sshd_config

	# Change the configuration of sshd, so that it does not automatically close
	# after the client has been idle for a while, it will now only disconnect
	# after 2 hours of inactivity.
	# - Allow PubKey Authentication
	# Reference: https://www.simplified.guide/ssh/disable-timeout
	# (sed) -i: Replace in-place, and append commands with -e
	# replace everything after and before the pattern with .*
	# sed -i '' -e 's/^#TCPKeepAlive.*/TCPKeepAlive no/' -e 's/^#ClientAliveInterval.*/ClientAliveInterval 30/' -e 's/^#ClientAliveCountMax.*/ClientAliveCountMax 240/' -e 's/^#PubkeyAuthentication.*/PubkeyAuthentication yes/' ./sshd_config	
	# Disable SSH timeout from the server. The server will not send any 
	# TCPKeepAlive packages and will only disconnect the client after 2 hours 
	# of inactivity.
	# Reference: https://www.simplified.guide/ssh/disable-timeout

	# Disable password authentication
	# In the default file from Vultr, there is an uncommented instance of 
	# PasswordAuthentication
	# sed -i '' -e 's/^PasswordAuthentication.*/PasswordAuthentication no/' -e 's/^#PasswordAuthentication.*/PasswordAuthentication no/' -e 's/^#ChallengeResponseAuthentication.*/ChallengeResponseAuthentication no/' ./sshd_config
	# If it is commented out
	# sed -i '' -e 's/^#PasswordAuthentication.*/PasswordAuthentication no/' ./sshd_config

	# Restart sshd with new configuration
	service sshd restart

}

# Connect through ssh to the remote VM and run all the commands there
connectVM(){
	# check if file with ssh secrets for VM exists, if so source file, transfer
    # script to VM with scp, and run the transferred script inside the VM after 	
	# establishing an ssh connection
	# if there is no file with credentials, return, since we are probably 
	# running the script inside the VM

	FILES_TO_SEND="${FILE_NAME} ${SSHD_CONFIG_FILE} ${FILE_USER_CONFIG}"
	# Use the "StrictHostKeyChecking no" ssh option, to not have to agree with
	# the key fingerprints of the server we are trying to ssh into
	[ -f ${FILE_VM_CREDENTIALS} ] && { source ${FILE_VM_CREDENTIALS}; scp -o "StrictHostKeyChecking no" ${FILES_TO_SEND} ${USER}@${HOST}:${PATH_IN_VM} && ssh ${USER}@${HOST} ${PATH_IN_VM}$(basename ${FILE_NAME}) && exit 0; } || return
}

main(){
	connectVM
	check_os
	install_bare_packages	
	#install_webdev_packages
	configure_dotfiles
	configure_ssh
}

main 


vultrClient/cmd/api/helpers.go
package main

import (
	"fmt"
	"net/http"
)

// addAuthToken, adds the necessary headers to get authenticated by the
// Vultr API.
func (app *application) addAuthToken(req *http.Request) {
	// Format token value for header. Add 'Bearer' before token
	tokenValue := fmt.Sprint("Bearer ", app.cfg.tokenAPI)
	req.Header.Add("Authorization", tokenValue)
	req.Header.Add("Content-Type", "application/json")
}

// checkResponseAPI, checks if the response from the API is correct, if not
// it returns error.
// The input parameter is the expected correct HTTP response (int) and the
// http.Response being checked.
func checkResponseAPI(resp *http.Response, correctResponse int) (err error) {
	// if everything went well Vultr API responds with "correctResponse" (int)
	if resp.StatusCode != correctResponse {
		if resp.StatusCode == 401 {
			err = fmt.Errorf("request failed. Response status: %s, expected %d %s. A 401 error response normally takes place, because the IP from which the request was done is not green-listed to communicate with the Vultr API. Check under 'Account/API/Access Control' in the Vultr control panel, if you need to enable access to the API for further IPv4 and IPv6 addresses to perform your next request.", resp.Status, correctResponse, http.StatusText(correctResponse))
			return
		}
		err = fmt.Errorf("request failed. Response status: %s, expected %d", resp.Status, correctResponse)
		return
	}
	return nil
}


vultrClient/cmd/api/apiModels.go
// This package contains the types to handle the responses of the Vultr API
package main

import (
	"time"
)

// RequestCreateInstance has the info necessary to request the creation of a
// new instance. If a value has the omitempty tag, then it is not strictly
// required to request the creation of a new instance.
type RequestCreateInstance struct {
	OS_ID    int    `json:"os_id"`
	Label    string `json:"label,omitempty"`
	Hostname string `json:"hostname,omitempty"`
	Region   string `json:"region"`
	Plan     string `json:"plan"`
	// Backups [enabled|disabled] automatic backups for the new instance.
	// Enabling automatic backups makes the instance more expensive..
	Backups string `json:"backups,omitempty"`
	// SSHKeys is a slice with the IDs of the SSH keys that should be used per
	// default when initializing the newly requested instance
	SSHKeys []string `json:"sshkey_id,omitempty"`
	// with backups is more expensive
}

// ResponseCreateInstance has the info received from response after requesting
// to create an instance
type ResponseCreateInstance struct {
	Instance *CreatedInstance `json:"instance"`
}

// CreatedInstance is the data structure that describes a created instance
type CreatedInstance struct {
	ID        string `json:"id"`
	OS        string `json:"os"`
	RAM       int    `json:"ram"`
	Disk      int    `json:"disk"`
	MainIP    string `json:"main_ip"`
	VCPUCount int    `json:"vcpu_count"`
	// Region is the region where the instance is being deployed
	Region           string    `json:"region"`
	Plan             string    `json:"plan"`
	CreationDate     time.Time `json:"date_created"`
	Status           string    `json:"status"`
	AllowedBandwidth int       `json:"allowed_bandwidth"`
	NetmaskV4        string    `json:"netmask_v4"`
	GatewayV4        string    `json:"gateway_v4"`
	PowerStatus      string    `json:"power_status"`
	ServerStatus     string    `json:"server_status"`
	V6Network        string    `json:"v6_network"`
	V6MainIP         string    `json:"v6_main_ip"`
	V6NetworkSize    int       `json:"v6_network_size"`
	Label            string    `json:"label"`
	InternalIP       string    `json:"internal_ip"`
	KVM              string    `json:"kvm"`
	Hostname         string    `json:"hostname"`
	Tag              string    `json:"tag"`
	OSID             int       `json:"os_id"`
	AppID            int       `json:"app_id"`
	ImageID          string    `json:"image_id"`
	FirewallGroupID  string    `json:"firewall_group_id"`
	Features         []string  `json:"features"`
	DefaultPassword  string    `json:"default_password"`
}

// ResponseLiveInstance has the info received from response after requesting
// the information from a live instance. The JSON response encapsulates the
// actual instance's data inside a JSON object first.
type ResponseLiveInstance struct {
	Instance *LiveInstance `json:"instance"`
}

// LiveInstance is the data structure that describes a live instance
type LiveInstance struct {
	ID        string `json:"id"`
	OS        string `json:"os"`
	RAM       int    `json:"ram"`
	Disk      int    `json:"disk"`
	MainIP    string `json:"main_ip"`
	VCPUCount int    `json:"vcpu_count"`
	// Region is the region where the instance is being deployed
	Region           string    `json:"region"`
	Plan             string    `json:"plan"`
	CreationDate     time.Time `json:"date_created"`
	Status           string    `json:"status"`
	AllowedBandwidth int       `json:"allowed_bandwidth"`
	NetmaskV4        string    `json:"netmask_v4"`
	GatewayV4        string    `json:"gateway_v4"`
	PowerStatus      string    `json:"power_status"`
	ServerStatus     string    `json:"server_status"`
	V6Network        string    `json:"v6_network"`
	V6MainIP         string    `json:"v6_main_ip"`
	V6NetworkSize    int       `json:"v6_network_size"`
	Label            string    `json:"label"`
	InternalIP       string    `json:"internal_ip"`
	KVM              string    `json:"kvm"`
	Hostname         string    `json:"hostname"`
	Tag              string    `json:"tag"`
	OSID             int       `json:"os_id"`
	AppID            int       `json:"app_id"`
	ImageID          string    `json:"image_id"`
	FirewallGroupID  string    `json:"firewall_group_id"`
	Features         []string  `json:"features"`
}

// SSHKey is the data structure with the fields that describe a single SSH key
type SSHKey struct {
	ID           string    `json:"id"`
	CreationDate time.Time `json:"date_created"`
	// Name is the name given to a single SSH key
	Name string `json:"name"`
	// SSH_Pub_Key is the actual public ssh key
	SSH_Pub_Key string `json:"ssh_key"`
}

// ResponseListSSHKeys has the information received after requesting a list with
// the stored SSH keys in the system
type ResponseListSSHKeys struct {
	SSHKeys []SSHKey `json:"ssh_keys"` // slice of ssh keys
	Meta    struct {
		TotalKeys int `json:"total"` // total amount of ssh keys
		Links     struct {
			Next string `json:"next"`
			Prev string `json:prev"`
		}
	}
}


vultrClient/cmd/api/apiMethods.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
)

// createInstance, create an Instance with the given plan in the specified
// regions, e.g. region="ewr" (New Jersey), plan="vc2-1c-1gb".
// Additionally, use a label and a hostname for the new instance, all this
// information is contained in the fields of the newInstance parameter.
func (app *application) createInstance(newInstance *RequestCreateInstance) (*CreatedInstance, error) {
	// Create a buffer with Read/Write methods implemented
	buf := new(bytes.Buffer)
	// Encode the data to JSON
	err := json.NewEncoder(buf).Encode(newInstance)
	if err != nil {
		return nil, err
	}

	URLInstancesVultrAPI := "https://api.vultr.com/v2/instances"
	req, err := http.NewRequest("POST", URLInstancesVultrAPI, buf)
	if err != nil {
		return nil, err
	}

	app.addAuthToken(req)

	// Send request to Vultr API
	resp, err := app.client.Do(req)
	defer resp.Body.Close()
	if err != nil {
		return nil, err
	}

	// if everything went well Vultr API responds with 202 (Status Accepted)
	if err = checkResponseAPI(resp, http.StatusAccepted); err != nil {
		return nil, err
	}

	responseJSON := new(ResponseCreateInstance)
	// decode response into JSON
	err = json.NewDecoder(resp.Body).Decode(responseJSON)
	if err != nil {
		return nil, err
	}
	// since the json response encapsulates the actual instance information,
	// inside an instance object, we have to return only this object
	return responseJSON.Instance, nil
}

// getInstance, fetch information of instance already deployed in the Vultr
// cloud. A pointer to a createdInstance as a parameter has all the information
// required to perform API call.
func (app *application) getInstance(instance *CreatedInstance) (*LiveInstance, error) {
	// Create/format string with URL for running instance
	URLGetInstanceVultrAPI := fmt.Sprintf("https://api.vultr.com/v2/instances/%s", instance.ID)
	req, err := http.NewRequest("GET", URLGetInstanceVultrAPI, nil)
	if err != nil {
		return nil, err
	}

	app.addAuthToken(req)

	// Send request to Vultr API
	resp, err := app.client.Do(req)
	defer resp.Body.Close()
	if err != nil {
		return nil, err
	}

	// if everything went well Vultr API responds with 200 (OK)
	if err = checkResponseAPI(resp, http.StatusOK); err != nil {
		return nil, err
	}

	responseJSON := new(ResponseLiveInstance)
	// decode response into JSON
	err = json.NewDecoder(resp.Body).Decode(responseJSON)
	if err != nil {
		return nil, err
	}
	// since the json response encapsulates the actual instance information,
	// inside an instance object, we have to return only this object
	return responseJSON.Instance, nil

}

// deleteInstance, deletes an actively running instance. The input parameter
// is a string with the instance's ID
func (app *application) deleteInstance(instanceID string) error {
	// Create/format string with URL for running instance
	URLDeleteInstanceVultrAPI := fmt.Sprintf("https://api.vultr.com/v2/instances/%s", instanceID)
	req, err := http.NewRequest("DELETE", URLDeleteInstanceVultrAPI, nil)
	if err != nil {
		return err
	}

	app.addAuthToken(req)

	// Send request to Vultr API
	resp, err := app.client.Do(req)
	defer resp.Body.Close()
	if err != nil {
		return err
	}

	// if everything went well Vultr API responds with 204 (No Content)
	// more information regarding this behaviour can be found here:
	// https://datatracker.ietf.org/doc/draft-ietf-httpbis-semantics/
	// If a DELETE method is successfully applied, the origin server SHOULD send
	//	*  a 202 (Accepted) status code if the action will likely succeed but
	//	has not yet been enacted,
	//
	//	*  a 204 (No Content) status code if the action has been enacted and
	//	no further information is to be supplied, or
	//
	//	*  a 200 (OK) status code if the action has been enacted and the
	//	response message includes a representation describing the status.
	// 	[...] The _204 (No Content)_ status code indicates that the server has
	// 	successfully fulfilled the request and that there is no additional
	//	content to send in the response content.
	if err = checkResponseAPI(resp, http.StatusNoContent); err != nil {
		return err
	}

	return nil

}

// listSSHKeys, list all SSH Keys registered to a particular Vultr account
func (app *application) listSSHKeys() ([]SSHKey, error) {
	URLSSHKeysVultrAPI := "https://api.vultr.com/v2/ssh-keys"
	req, err := http.NewRequest("GET", URLSSHKeysVultrAPI, nil)
	app.addAuthToken(req)

	// Send request to Vultr API
	resp, err := app.client.Do(req)
	defer resp.Body.Close()
	if err != nil {
		return nil, err
	}
	// if everything went well VultrAPI respondes with 200 (OK)
	if err = checkResponseAPI(resp, http.StatusOK); err != nil {
		return nil, err
	}

	responseJson := new(ResponseListSSHKeys)
	err = json.NewDecoder(resp.Body).Decode(responseJson)
	if err != nil {
		return nil, err
	}

	return responseJson.SSHKeys, nil
}


vultrClient/cmd/api/main.go
// Eduardo Rodriguez @erodrigufer (c) 2022

package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"
)

// application type used for dependency injection / avoid using globals
type application struct {
	// errorLog, error log handler
	errorLog *log.Logger
	// infoLog, info log handler
	infoLog *log.Logger
	// cfg, config values (e.g. from flags)
	cfg *configValues
	// client, long-living http client used to communicate with API
	client *http.Client
}

// store all flag-parseable config values in this struct
type configValues struct {
	// tokenAPI, API token used to communicate with Vultr API
	tokenAPI string
	// sshkey, ssh key which will be automatically initialized in new instance
	sshkey string
	// hostname, hostname for new instance
	hostname string
	// label, label for new instance
	label string
	// region, region where to deplay VM
	region string
	// plan, the kind of VM that will be deployed
	plan string
	//osID, the OS do be installed in the new instance
	osID int
	// action to perform with this script, e.g. 'deploy', 'list', 'delete'
	action string
	// perform action on this instance (related through ID)
	instanceID string
}

func main() {
	// Create a logger for INFO messages, the prefix "INFO" and a tab will be
	// displayed before each log message. The flags Ldate and Ltime provide the
	// local date and time
	infoLog := log.New(os.Stdout, "INFO\t", log.Ldate|log.Ltime)

	// Create an ERROR messages logger, additionally use the Lshortfile flag to
	// display the file's name and line number for the error
	errorLog := log.New(os.Stderr, "ERROR\t", log.Ldate|log.Ltime|log.Lshortfile)

	app := &application{
		infoLog:  infoLog,
		errorLog: errorLog,
	}

	cfg := new(configValues)
	flag.StringVar(&cfg.tokenAPI, "tokenAPI", "", "Personal Access Token to interact with Vultr API")
	flag.StringVar(&cfg.sshkey, "sshKey", "", "SSH key to initialize per default in new instance")
	flag.StringVar(&cfg.hostname, "hostname", "", "Hostname for new instance")
	flag.StringVar(&cfg.label, "label", "", "Label for new instance")
	flag.StringVar(&cfg.region, "region", "fra", "Region where to deply VM")
	flag.StringVar(&cfg.action, "action", "", "Action to perform with this script")
	flag.StringVar(&cfg.instanceID, "instanceID", "", "Actions are performed on this instance (ID)")
	flag.StringVar(&cfg.plan, "plan", "vc2-1c-1gb", "Create an instance with this plan")
	flag.IntVar(&cfg.osID, "os", 447, "OS of the new instance (default FreeBSD-13)")
	flag.Parse()
	if cfg.tokenAPI == "" {
		app.errorLog.Fatal("missing API token")
	}
	if cfg.action == "" {
		// TODO: print usage and option for action
		app.errorLog.Fatal("no action was specified")
	}

	app.cfg = cfg

	// Use a single http.Client for all interactions with the API, as stated in
	// the official Go documentation "The Client's Transport typically has
	// internal state (cached TCP connections), so Clients should be reused
	// instead of created as needed. Clients are safe for concurrent use by
	// multiple goroutines."
	app.client = new(http.Client)

	newInstance := &RequestCreateInstance{
		OS_ID:   app.cfg.osID,
		Region:  app.cfg.region,
		Backups: "disabled",
		// Enabling backups makes the VM more expensive
		Plan:     app.cfg.plan,
		Hostname: app.cfg.hostname,
		Label:    app.cfg.label,
	}

	// Append ssh key parsed from flags, if no ssh key was parsed an empty
	// string is appended
	newInstance.SSHKeys = append(newInstance.SSHKeys, app.cfg.sshkey)

	// create a new instance
	if app.cfg.action == "create" {
		if err := app.actionCreate(newInstance); err != nil {
			app.errorLog.Println(err)
			os.Exit(1)
		}
		os.Exit(0)
	}

	// delete instance
	if app.cfg.action == "delete" {
		if err := app.actionDelete(app.cfg.instanceID); err != nil {
			app.errorLog.Println(err)
			os.Exit(1)
		}
		os.Exit(0)
	}
	//keys, err := app.listSSHKeys()
	//if err != nil {
	//	app.errorLog.Println(err)
	//	return
	//}
	//fmt.Printf("SSH Keys: %+v\n", keys)
}

// actionCreate, if the flag action is equal to create, a new instance will be
// created
func (app *application) actionCreate(newInstance *RequestCreateInstance) error {
	createdInstance, err := app.createInstance(newInstance)
	if err != nil {
		return err
	}
	app.infoLog.Printf("new instance [ID: %s] created.", createdInstance.ID)

	// Declare instance outside for-loop to be able to use it afterwards as well
	instance, _ := app.getInstance(createdInstance)

	// Ping newly created instance to get its IP address
	for i := 0; i < 200; i++ {
		time.Sleep(5 * time.Second)
		instance, err = app.getInstance(createdInstance)
		if err != nil {
			app.errorLog.Println(err)
			continue // try checking again, getInstance() method failed
		}
		if instance.Status == "active" {
			app.infoLog.Printf("Instance ID: %s. Status: %s. Main IP: %s", instance.ID, instance.Status, instance.MainIP)
			break // Status is now active, stop pinging the API
		}
	}

	if instance.MainIP == "" {
		return fmt.Errorf("The instance's main IP address could not be determined!")
	}

	// Store the IP address of the new instance in a text file
	data := []byte(fmt.Sprintf("USER=root\nHOST=%s\nINSTANCE_ID=%s\n", instance.MainIP, instance.ID))
	// Write IP address to output file, to better automate working with shell
	// script that automates installing all the packages.
	// Only owner has read priviledges on file.
	err = os.WriteFile("vm_credentials.secrets", data, 0400)
	if err != nil {
		return err
	}
	app.infoLog.Println("VM's credentials written to file")

	return nil

}

// actionDelete, if the flag action is equal to delete, the specified instance
// will be deleted. Parameter is the instance's ID as a string
func (app *application) actionDelete(instanceID string) error {
	if err := app.deleteInstance(instanceID); err != nil {
		return err
	}
	app.infoLog.Printf("Instance %s was correctly deleted.\n", instanceID)
	return nil
}
